src/nxt_h1proto.c

```diff
@@ -2353,6 +2362,61 @@ nxt_h1p_peer_header_send(nxt_task_t *task, nxt_http_peer_t *peer)
     }
 
     nxt_conn_write(task->thread->engine, c);
+
+    return;
+
+fail:
+
+    r->state->error_handler(task, r, peer);
+}
+
+
+static nxt_int_t
+nxt_h1p_peer_request_target(nxt_http_request_t *r, nxt_str_t *target)
+{
+    u_char  *p;
+    size_t  size, encode;
+
+    if (!r->uri_changed) {
+        *target = r->target;
+        return NXT_OK;
+    }
+
+    if (!r->quoted_target && r->args->length == 0) {
+        *target = *r->path;
+        return NXT_OK;
+    }
+
+    if (r->quoted_target) {
+        encode = nxt_encode_complex_uri(NULL, r->path->start,
+                                        r->path->length);
+    } else {
+        encode = 0;
+    }
+
+    size = r->path->length + encode * 2 + 1 + r->args->length;
+
```

Comment:
Just want to confirm that the required calculation is (taking operator precedence into account) is

```c
size = r->path->length + (encode * 2) + 1 + r->args->length;
```

e.g

```
5 + (7*2) + 1 + 8 = 28
```

Reply:
Yes, I did.

Reply:
Final question about this.

Why the * 2?

Reply:
According to its implementation.
https://github.com/nginx/unit/blob/master/src/nxt_string.c#L607
Take `/a b c` as an example, the allocation length is `r->path->length` + `encoded_number * 2`.
```
/a b c => /a%20b%20c
r->path: "/a b c"
encoded chars: ' ', ' '
total: 6 + 2 * 2
```

Reply:
Got it.

r->path->length already includes the to be encoded chars, and each to be encoded char adds two bytes.

Reply:
Correct.